python_modules/dagster/dagster/_core/op_concurrency_limits_counter.py

```diff
@@ -0,0 +1,120 @@
+from collections import defaultdict
+from typing import Optional, Sequence
+
+import pendulum
+
+from dagster._core.snap.execution_plan_snapshot import ExecutionPlanSnapshot
+
+from .instance import DagsterInstance
+from .storage.dagster_run import DagsterRun, DagsterRunStatus, RunRecord
+from .storage.tags import GLOBAL_CONCURRENCY_TAG
+
+
+def compute_root_concurrency_keys_for_snapshot(
+    plan_snapshot: ExecutionPlanSnapshot,
+) -> Sequence[Optional[str]]:
+    """Utility function called at run creation time to add the concurrency info needed to keep track
+    of concurrency limits for each in-flight run.
+    """
+    root_step_keys = set(
+        [step_key for step_key, deps in plan_snapshot.step_deps.items() if not deps]
+    )
+    return [
+        (step.tags or {}).get(GLOBAL_CONCURRENCY_TAG)
+        for step in plan_snapshot.steps
+        if step.key in root_step_keys
+    ]
```

Comment:
as discussed offline i think it would be good for this to not write a bunch of Nones on every run in the common case  where there are no keys

A map from keys|None to counts could also be nice, depending on how common we think it will be that different root nodes in the same run will have the same concurrency key

and finally we could track 'does have any nodes without a key' as a separate boolean (like make this a serdes object instead of just a list, with a boolean and the list)

It definitely was a journey for me to understand what duplicates in this list meant (and that there could even be Nones, let alone what they meant) - so making that clearer seems good, either via more comments or by adjusting the data structure to give it more named fields

Reply:
I initially changed this to be an optional dict with None or str => int, where the int was the count and the None|str was the concurrency key.

But this started to get into weird serialization errors, where things started deserializing into the string `null`.

So, now I maintain two fields on the run:
1. a map of root concurrency keys to counts
2. a boolean flag if there are any root ops that do not have concurrency keys